### PR TITLE
#19756: Add extractData and extractRequestEntity directives.

### PR DESCRIPTION
**Pull request by kwyczesany**
Add two additional extract functions to BasicDirective and tests for them. I'm not sure if the solution is correct so when the code will be ok, I'll add documentation.

related to #19756

### CODE REVIEW

**Comment by typesafehub-validator** [comment]
Hi @kwyczesany,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Typesafe Contributors License Agreement:

[http://www.lightbend.com/contribute/cla](http://www.lightbend.com/contribute/cla)

**Comment by ktoso** [comment]
OK TO TEST

File: akka-http-tests/src/test/scala/akka/http/scaladsl/server/directives/BasicDirectivesSpec.scala [hunk 1]
```diff
@@ -26,4 +30,27 @@ class BasicDirectivesSpec extends RoutingSpec {
       } ~> check { responseAs[String] shouldEqual "GET" }
     }
   }
+
+  "The `extractData` directive" should {
+    "extract stream of ByteString from the RequestContext" in {
+      val dataBytes = Source.fromIterator(() ⇒ Iterator.range(1, 10) map (x ⇒ ByteString(x + "")))
```
Comment (ktoso) [inline akka-http-tests/src/test/scala/akka/http/scaladsl/server/directives/BasicDirectivesSpec.scala]: `.map()` please :)

File: akka-http-tests/src/test/scala/akka/http/scaladsl/server/directives/BasicDirectivesSpec.scala [hunk 2]
```diff
@@ -26,4 +30,27 @@ class BasicDirectivesSpec extends RoutingSpec {
...
+      Post("/abc", HttpEntity(ContentTypes.`text/plain(UTF-8)`, data = dataBytes)) ~> {
+        extractData { data ⇒
```
Comment (ktoso) [inline akka-http-tests/src/test/scala/akka/http/scaladsl/server/directives/BasicDirectivesSpec.scala]: maybe `extractDataBytes` as it relates more naming wise to what we're doing?

Comment (ktoso) [inline akka-http-tests/src/test/scala/akka/http/scaladsl/server/directives/BasicDirectivesSpec.scala]: (data sounds a bit like form data)

File: akka-http/src/main/scala/akka/http/javadsl/server/directives/BasicDirectives.scala [hunk 3]
```diff
@@ -269,4 +271,16 @@ abstract class BasicDirectives {
     D.extractRequestContext { ctx ⇒ inner.apply(JavaMapping.toJava(ctx)(server.RoutingJavaMapping.RequestContext)).delegate }
   }
 
+  /**
+   * Extracts the [[akka.stream.javadsl.Source]] from the [[akka.http.javadsl.server.RequestContext]].
```
Comment (ktoso) [inline akka-http/src/main/scala/akka/http/javadsl/server/directives/BasicDirectives.scala]: `Extracts the entities dataBytes [[...Soruce] ...`

**Comment by ktoso** [comment]
Looks good! Just need to add docs with examples now :)

**Comment by johanandren** [comment]
I just +1:ed on every note by @ktoso, fix those and add docs. Looks promising!

**Comment by drewhk** [comment]
LGTM

**Comment by ktoso** [comment]
LGTM, great work! :-)

**Pull request merged**